JDK-8250768: javac should be adapted to changes in JEP 12

lahodaj opened the pull request:
This is an update to javac and javadoc, to introduce support for Preview APIs, and generally improve javac and javadoc behavior to more closely adhere to JEP 12.

The notable changes are:

 * adding support for Preview APIs (javac until now supported primarily only preview language features, and APIs associated with preview language features). This includes:
     * the @PreviewFeature annotation has boolean attribute "reflective", which should be true for reflective Preview APIs, false otherwise. This replaces the existing "essentialAPI" attribute with roughly inverted meaning.
     * the preview warnings for preview APIs are auto-suppressed as described in the JEP 12. E.g. the module that declares the preview API is free to use it without warnings
 * improving error/warning messages. Please see [1] for a list of cases/examples.
 * class files are only marked as preview if they are using a preview feature. [1] also shows if a class file is marked as preview or not.
 * the PreviewFeature annotation has been moved to jdk.internal.javac package (originally was in the jdk.internal package).
 * Preview API in JDK's javadoc no longer needs to specify @preview tag in the source files. javadoc will auto-generate a note for @PreviewFeature elements, see e.g. [2] and [3] (non-reflective and reflective API, respectively). A summary of preview elements is also provided [4]. Existing uses of @preview have been updated/removed.
 * non-JDK javadoc is also enhanced to auto-generate preview notes for uses of Preview elements, and for declaring elements using preview language features [5].
 
 Please also see the CSR [6] for more information.
 
 [1] http://cr.openjdk.java.net/~jlahoda/8250768/JEP12-errors-warnings-v6.html
 [2] http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.00/api/java.base/java/lang/Record.html
 [3] http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.00/api/java.compiler/javax/lang/model/element/RecordComponentElement.html
 [4] http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.00/api/preview-list.html
 [5] http://cr.openjdk.java.net/~jlahoda/8250768/test.javadoc.00/
 [6] https://bugs.openjdk.java.net/browse/JDK-8250769

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8250768](https://bugs.openjdk.java.net/browse/JDK-8250768): javac should be adapted to changes in JEP 12


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to be1d8651cbb45bd3bcab168651abef09372580c1
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to efb37a9b7eb63ce252ce86a518154e1f3da833d1
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to a8046dde96b555c1c846496cfdd7ed8720fb7cfa
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 0c1c4d57c293db23badee1e00265e11f3e3d09be


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/703/head:pull/703`
`$ git checkout pull/703`

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:
```diff
@@ -2646,6 +2646,7 @@ JCStatement parseStatementAsBlock() {
             }
         }
         if (isRecordStart() && allowRecords) {
+            checkSourceLevel(Feature.RECORDS);
```
Is this change really related to the JEP 12?

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:
```diff
@@ -3709,6 +3710,7 @@ protected JCStatement classOrRecordOrInterfaceOrEnumDeclaration(JCModifiers mods
         if (token.kind == CLASS) {
             return classDeclaration(mods, dc);
         } if (isRecordStart()) {
+             checkSourceLevel(Feature.RECORDS);
```
Same here

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/util/RawDiagnosticFormatter.java:
```diff
@@ -160,11 +162,17 @@ protected String formatArgument(JCDiagnostic diag, Object arg, Locale l) {
         } else if (arg instanceof Tag) {
             s = "compiler.misc.tree.tag." + StringUtils.toLowerCase(((Tag) arg).name());
+        } else if (arg instanceof Source && arg == Source.DEFAULT &&
+                CODES_NEEDING_SOURCE_NORMALIZATION.contains(diag.getCode())) {
```
Nice trick to keep raw output constant across versions :-)

mcimadamore reviewed (commented):
I've done a quick pass over the javac changes, things look good, apart from some (minor) comments.

I'm a bit worried about the scalability of the output in this page:

http://cr.openjdk.java.net/~jlahoda/8250768/test.javadoc.00/use/Factory.html

E.g. if the types which use preview features are a lot (e.g. 50), what would the text inside the black box look like?

erikj79 reviewed (approved):
Build changes look good.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:
```diff
@@ -197,7 +198,23 @@ public void addClassSignature(String modifiers, Content classInfoTree) {
+            if ("record".equals(modifiersPart) ||
+                "sealed".equals(modifiersPart) || "non-sealed".equals(modifiersPart)) {
+                pre.add(modifiersPart);
+                pre.add(new HtmlTree(TagName.SUP).add(HtmlTree.A("#preview", contents.previewMark)));
```
This will likely clash with a public field named "preview". You should choose a name for the anchor that is not a Java identifier.

hns commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -2151,4 +2213,152 @@ Script getMainBodyScript() {
+            DocTree previewTree = utils.getPreviewTree(forWhat);
+            if (previewTree != null) {
+                previewDiv.add(new HtmlTree(TagName.A).put(HtmlAttr.ID, "preview")
+                                                      .add(new RawHtml(utils.getPreviewTreeSummaryOrDetails(previewTree, false))));
```
The `id` attribute needs to be unique within the page, so in addition to make the value not a valid java identifier (as @jonathan-gibbons pointed out in a comment elsewhere) we need to support multiple preview ids per page. One way to do this would be to add the element name to the id value, e.g. `preview-<element name>`.

hns replied:
Of course the element name won't do for overloaded methods and constructors... `Links#getAnchor(ExecutableElement)` should be used for those.

lahodaj replied:
Uh, originally, there was only preview section per file, and I didn't fully realize the JDK's javadoc may have multiple such section. I'll work on this. Thanks for the pointer!

lahodaj commented:
I've updated the patch based on the comments: mostly updating the anchors in the javadoc, but also removing the updates to JavacParser, which are only loosely related to the primary focus of this patch, and should possibly be done separately.

mcimadamore reviewed (approved):
javac changes look good

hns reviewed (commented):
The javadoc code changes look reasonable, and the preview bits in the generated documentation look good as well. 

Apart from my inline comments which all address minor issues, there are a few things I don't like around `PreviewListWriter`: its code replication with `DeprecatedListWriter, and it adds things to HtmlDocletWriter and HtmlStyle that I don't think are strictly necessary. However, I wouldn't expect you to know how we like things done in javadoc, so maybe the simplest solution would be for one of us to go over the javadoc bits, either as part of this pull request or (if you prefer to get it integrated rather sooner) as a follow-up task.

lahodaj commented:
I have update the patch based on Hannes' comments. One open question is whether we should have Utils.elementFlags, or just Utils.isPreviewElement. @jonathan-gibbons, any preference? Thanks!

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2924,4 +2947,237 @@ public String toString() {
+        return isPreviewAPI(el) && configuration.workArounds.isReflectivePreviewAPI(el);
+    }
+
+    public Set<ElementFlag> elementFlags(Element el) {
```
There's a number of predicates on an element that the doclet might be interested in that could be cached/reified as "flags". Today, we have "preview" and "deprecated" that have similar handling; there have been discussions about handling native methods in a similar fashion.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java:
```diff
@@ -33,13 +33,19 @@
 import javax.lang.model.element.TypeElement;
 
 import com.sun.source.doctree.DocTree;
+import com.sun.source.doctree.InlineTagTree;
+import com.sun.source.doctree.TextTree;
+import com.sun.source.doctree.UnknownInlineTagTree;
+import java.util.stream.Collectors;
```
why these imports?

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:
```diff
@@ -270,6 +274,18 @@ doclet.navClassUse=Use
 doclet.Error_in_grouplist=Bad -group option: {0} {1}
 doclet.Groupname_already_used=In -group option, group name already used: {0}
 doclet.Same_element_name_used=Element name or pattern used twice: {0}
+doclet.PreviewLeadingNote=<code>{0}</code> relies on preview features of the Java platform:
+doclet.Declared_Using_Preview=<code>{0}</code> is declared using <em>{1}</em>, a preview feature of the Java language ({2}).
+doclet.PreviewAPI=<code>{0}</code> refers to one or more preview APIs:  {1}.
+doclet.ReflectivePreviewAPI=<code>{0}</code> refers to one or more reflective preview APIs:  {1}.
+doclet.UsesDeclaredUsingPreview=<code>{0}</code> refers to one or more types which are declared using a preview feature of the Java language: {1}.
+doclet.PreviewTrailingNote1=Programs can only use <code>{0}</code> when preview features are enabled.
+doclet.PreviewTrailingNote2=Preview features may be removed in a future release, or upgraded to permanent features of the Java platform.
+doclet.Declared_Using_Preview.RECORD=Records
+doclet.Declared_Using_Preview.SEALED=Sealed Classes
+doclet.Declared_Using_Preview.SEALED_PERMITS=Sealed Classes
+doclet.PreviewPlatformLeadingNote=<code>{0}</code> is a preview API of the Java platform.
+doclet.ReflectivePreviewPlatformLeadingNote=<code>{0}</code> is a reflective preview API of the Java platform.
```
HTML in resource files is bad.  It would be marginally better to move the HTML to the value being substituted (using strings from Content nodes) and even better to use a method that substitutes Content nodes into a resource string (Not sure if that method exists, but it could/should).

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/ConstructorWriter.java:
```diff
@@ -70,6 +70,14 @@
+     * @param member the member being documented
+     * @param annotationDocTree content tree to which the preview information will be added
+     */
+    void addPreview(ExecutableElement member, Content contentTree);
```
Note to javadoc-team:  Consider using default methods to provide an empty implementation.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/builders/ConstructorBuilder.java:
```diff
@@ -25,11 +25,13 @@
 
 package jdk.javadoc.internal.doclets.toolkit.builders;
 
+import static java.lang.invoke.ConstantBootstraps.enumConstant;
```
Really? If it is required, it is in the wrong place.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -0,0 +1,184 @@
+ *  This code and its internal interfaces are subject to change or
+ *  deletion without notice.</b>
+ */
+public class PreviewAPIListBuilder {
```
OK for now, but it might be worth eventually trying to merge this with `DeprecatedListBuilder`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -0,0 +1,184 @@
+        previewMap = new EnumMap<>(PreviewElementKind.class);
+        for (PreviewElementKind kind : PreviewElementKind.values()) {
+            previewMap.put(kind,
+                    new TreeSet<>(utils.comparators.makeDeprecatedComparator()));
```
The use of `makeDeprecatedComparator` is not awful, but does smell a bit.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    /**
+     * Build the sorted list of all the deprecated APIs in this run.
+     * Build separate lists for deprecated modules, packages, classes, constructors,
```
"d-word", twice

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -0,0 +1,184 @@
+     * Build separate lists for deprecated modules, packages, classes, constructors,
+     * methods and fields.
+     */
+    private void buildDeprecatedAPIInfo() {
```
D-word

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -0,0 +1,184 @@
+                        break;
+                }
+            }
+            composeDeprecatedList(previewMap.get(PreviewElementKind.FIELD),
```
I suggest you grep this file for all uses of the d-word

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -0,0 +1,184 @@
+     * @param sset set of deprecated elements.
+     * @param members members to be added in the list.
+     */
+    private void composeDeprecatedList(SortedSet<Element> sset, List<? extends Element> members) {
```
Last time d-word comment. Consider all such uses to be flagged.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2831,6 +2834,26 @@ public DocCommentTree getDocCommentTree(Element element) {
+    public String getPreviewTreeSummaryOrDetails(DocTree t, boolean summary) {
+        UnknownInlineTagTree previewTag = (UnknownInlineTagTree) t;
+        List<? extends DocTree> previewContent = previewTag.getContent();
+        String previewText = ((TextTree) previewContent.get(0)).getBody();
```
This looks unreasonably fragile. And, I thought the tag was going away... at least according to earlier files in this review.

lahodaj replied:
This was intended to allow Preview APIs to provide their own text instead of the generic one. But, looking again at JEP 12, this shouldn't be supported, so removing.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2924,4 +2947,237 @@ public String toString() {
+
+        SEALED(List.of("sealed")),
+        SEALED_PERMITS(List.of("sealed", "permits")),
+        RECORD(List.of("record"));
```
I'm guessing this is about to go away soon?

lahodaj replied:
Right, this is likely to go away soon.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2924,4 +2947,237 @@ public String toString() {
+        return result;
+    }
+
+    public enum DeclarationPreviewLanguageFeatures {
```
General thinking aloud question ... how does all this interact with source or release options for an earlier release?

lahodaj replied:
I don't think there should be much interaction with -source <older-version>. We don't support preview features from previous version or preview class files from previous versions, so I think it should be enough to handle the currently present preview features.

lahodaj replied:
We don't support preview features from previous releases, AFAIK. So javadoc for JDK 16 should not accept e.g. record class when running with  -source 15.

jonathan-gibbons replied:
Yeah, my recollection is that I was wondering whether preview-related code needs to be "guarded" to only work in the current release. But, I guess we may get the right effect (of forbidding preview features in older code) from the javac front end, so that in javadoc we can be assured that there are no instances of what may still be preview features in older code (i.e with older --source/--rlease options)

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2924,4 +2947,237 @@ public String toString() {
+            }
+            case MODULE, PACKAGE -> {
+            }
+            default -> throw new IllegalStateException("Unexpected: " + el.getKind());
```
Should be `IllegalArgumentException` not `IllegalStateException`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2924,4 +2947,237 @@ public String toString() {
+    /**
+     * Checks whether the given Element should be marked as a preview API.
+     *
+     * Note that is a type is marked as a preview, its members are not.
```
probable typo: "is" -> "if"

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/ElementsTable.java:
```diff
@@ -1285,6 +1285,7 @@ private ElementKind getAllowedKind(ElementKind kind) {
                 case CONSTRUCTOR: case ENUM_CONSTANT: case EXCEPTION_PARAMETER:
                 case FIELD: case INSTANCE_INIT: case LOCAL_VARIABLE: case PARAMETER:
                 case RESOURCE_VARIABLE: case STATIC_INIT: case TYPE_PARAMETER:
+                case RECORD_COMPONENT:
```
I'm not saying this is wrong, but I'd like to understand why it is necessary.

lahodaj replied:
HtmlDocletWriter.getPreviewNotes analyzes classes and their members, to see if some are using aspects that are under preview. When looking at the members, it uses utils.isIncluded on the member, and that eventually gets here. And if the member is a RECORD_COMPONENT, it would fail here. But we can avoid asking for RECORD_COMPONENTS, if needed.

jonathan-gibbons replied:
ok

jonathan-gibbons reviewed (commented):
I have primarily gone through all the javadoc changes.

There are three main areas of feedback:

1. The ElementFlag question. We have gone from one kind of element with special treatment (deprecated) to two (deprecated, preview), and there are signs of a third kind coming soon, maybe as early as next year, for work currently being discussed in the Panama project.

As the saying goes, three would be one too many.

So, I agree `ElementFlag` is underutilized today and could reasonably be removed, but it does seem worthwhile to keep, and it would even be worth increasing its usage soon, such as to combine methods and classes for deprecated elements and preview elements. 

I'm not sure I can go back into looking at files while typing this message, but if we are to keep `ElementFlag` we should at a minimum provide a better description of its purpose. For example, can/should it be used for all predicates on elements, or just the elements that get "special" handling when generating docs.

2. The use of strings containing HTML, and use of `RawHtml`, instead of working in terms of `Content` nodes such as `HtmlTree` and `StringContent`.

3. Track recent updates to the repo, regarding Conditional Pages. See how we set up conditional pages for the deprecated list, and do the same for preview items.  This is probably a must-do item, once you merge with mainline.
--

Finally, I realize this is a big chunk of work (well done!), across many areas, and that getting through a review is hard.  If it is too hard to address some of the comments here, I'm OK if you file follow-up bugs of at least the same priority and Fix Version as for the main bug here: JDK-8250768. (That applies to #1, #2 above; not #3).

magicus reviewed (approved):
Build changes look good.

lahodaj commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractExecutableMemberWriter.java:
```diff
@@ -84,7 +85,7 @@ protected Content getTypeParameters(ExecutableElement member) {
 
     @Override
-    protected Content getDeprecatedLink(Element member) {
+    protected Content getDeprecatedOrPreviewLink(Element member) {
```
Note this method returns the same link for deprecate or preview - it just was named "getDeprecatedLink", and when using it work preview, I renamed it "getDeprecatedOrPreviewLink". We may need to think of a better name.

lahodaj commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Navigation.java:
```diff
@@ -223,6 +225,7 @@ private void addMainNavLinks(Content tree) {
                 addPageLabel(tree, contents.useLabel, options.classUse());
                 addTreeLink(tree);
                 addDeprecatedLink(tree);
+                addPreviewLink(tree);
```
I believe the link is in the navigation bar on all pages (as DEPRECATED is).

jonathan-gibbons commented on test/langtools/tools/javac/patterns/BindingsTest2.out:
```diff
@@ -49,6 +49,6 @@ BindingsTest2.java:241:17: compiler.err.cant.resolve.location: kindname.variable
-- compiler.note.preview.filename: BindingsTest2.java
+- compiler.note.preview.filename: BindingsTest2.java, DEFAULT
 - compiler.note.preview.recompile
 51 errors
```
(issue unrelated to review)  it seems wrong that we encourage/allow the use of files that do not end in newline.

jonathan-gibbons commented on test/langtools/tools/javac/preview/PreviewAutoSuppress.java:
```diff
@@ -0,0 +1,221 @@
+ */
+
+/*
+ * @test
```
no `@bug`?

jonathan-gibbons commented on test/langtools/tools/javac/preview/PreviewErrors.java:
```diff
@@ -27,15 +27,16 @@
  * @build combo.ComboTestHelper
- * @compile --enable-preview -source ${jdk.version} PreviewErrors.java
- * @run main/othervm --enable-preview PreviewErrors
+ * @compile PreviewErrors.java
```
do you need `@compile` here: is `@build` not good enough?

jonathan-gibbons reviewed (commented):
I have read all the files. 

I have added a n umber of various minor non-blocking comments (no need for re-review( to fix these.  But I have a couple of comments/questions before finally giving approval.
There's a comment in `PreviewListWriter` about annotation members that needs too be addressed, and I wonder is RECORD and RECORD_COMPONENT need to be added into PreviewElementKind.

lahodaj commented on src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacTaskPool.java:
```diff
@@ -251,6 +252,9 @@ public void printStatistics(PrintStream out) {
         }
 
         void clear() {
+            //when patching modules (esp. java.base), it may be impossible to
+            //clear the symbols read from the patch path:
+            polluted |= get(JavaFileManager.class).hasLocation(StandardLocation.PATCH_MODULE_PATH);
```
The tests need to inject preview APIs into compilation, and without this tweak, the incorrect values are re-used for tests that use the javac pool. So it is related in the sense it is needed for the tests to pass.

lahodaj commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/builders/AnnotationTypeRequiredMemberBuilder.java:
```diff
@@ -152,6 +153,7 @@ protected void buildAnnotationTypeMember(Content detailsList)
     protected void buildAnnotationTypeMemberChildren(Content annotationDocTree) {
         buildSignature(annotationDocTree);
         buildDeprecationInfo(annotationDocTree);
+        buildPreviewInfo(annotationDocTree);
```
Right, I think so.

lahodaj commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/doclets.properties:
```diff
@@ -208,6 +208,7 @@ doclet.Use_Table_Summary=Use table, listing {0}, and an explanation
 doclet.Member_Table_Summary={0} table, listing {1}, and an explanation
 doclet.fields=fields
 doclet.Fields=Fields
+doclet.Preview=Preview.
 doclet.Properties=Properties
 doclet.constructors=constructors
```
This is consistent with doclet.Deprecated key a few lines up. In the javadoc it then says something along these lines:
```
Preview.
A binding pattern tree
```

lahodaj commented on test/langtools/jdk/javadoc/doclet/testPreview/api/preview/Core.java:
```diff
@@ -0,0 +1,30 @@
+import jdk.internal.javac.PreviewFeature;
+import jdk.internal.javac.PreviewFeature.Feature;
+
+@PreviewFeature(feature=Feature.TEST)
```
I have intentionally added Feature.TEST to improve testability. Before, tests were using one of the constants (typically whatever was the first constant), but that seems somewhat problematic - what if (at some point, transiently) we have no preview features?

lahodaj commented:
Thanks @jonathan-gibbons for your comments! I've tried to update the code based on them, mostly in https://github.com/lahodaj/jdk/commit/743f516c660b577035cdda4510a0bb97937fd9b2 and https://github.com/lahodaj/jdk/commit/e4b02827998fc2e8f19f983aabfb3d720b03d111

A big chunk of the update is generalization of the deprecated and preview list builders and writers into a "summary" list builder and writer. These should also now handle records. For record components, those are a little tricky, as (AFAIK) can't currently have deprecation/preview-ness for them (and hence there is no good way to test any support for record components in these). But the summary build and writer are looking for record components and will fail in case a record component is sent into them.

lahodaj commented:
FWIW, a javadoc generated with the current version of the patch:
http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.01/api/index.html

And a specdiff comparing it to the javadoc built from the corresponding master:
http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.specdiff.01/overview-summary.html

jonathan-gibbons commented:
The page and comments are generally good, but there is a bug that needs 
to be fixed.

After clicking on the PREVIEW link in the top bar to go to the Preview 
page, the word DEPRECATED is highlighted in the top navbar instead of 
PREVIEW.

-- Jon


On 11/5/20 10:13 AM, mlbridge[bot] wrote:
>
> /Mailing list message from Alex Buckley 
> <mailto:alex.buckley@oracle.com> on kulla-dev 
> <mailto:kulla-dev@openjdk.java.net>:/
>
> On 11/5/2020 4:45 AM, Jan Lahoda wrote:
>
>     FWIW, a javadoc generated with the current version of the patch:
>     http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.01/api/index.html
>
> Allow me to draw people's attention to the PREVIEW link in the banner of
> the generated javadoc. It shows all the preview APIs in the release on
> one page. This is very helpful for understanding the surface area of a
> preview feature.
>
> For example, with Sealed Classes being the only preview feature likely
> to target JDK 16, the PREVIEW page shows that the feature's API is
> solely about reflection. It's clear that Sealed Classes do not
> introduce, say, a java.lang.Sealed class analogous to the
> java.lang.Record class introduced by Records in JDK 14/15 (and which
> would have appeared on the PREVIEW page had it existed then).
>
> Alex
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://urldefense.com/v3/__https://github.com/openjdk/jdk/pull/703*issuecomment-722549484__;Iw!!GqivPVa7Brio!ONrE9EESNKSCZJPGXBa8TTa4Ey9OqEVTynHLfgVOrBRIkb1Cr2iqLALb01M_DlGDeQEPpw$>, 
> or unsubscribe 
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AOUXBRRTA7I6T2AUM3QT7QTSOLTLXANCNFSM4STPEYJQ__;!!GqivPVa7Brio!ONrE9EESNKSCZJPGXBa8TTa4Ey9OqEVTynHLfgVOrBRIkb1Cr2iqLALb01M_DlE4QxPnyw$>.
>

lahodaj commented:
Thanks for noticing Chris and @jonathan-gibbons. Fixed in 61264fe, updated javadoc:
http://cr.openjdk.java.net/~jlahoda/8250768/jdk.javadoc.02/api/preview-list.html

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -2942,4 +2944,256 @@ public String toString() {
     }
+
+    /**
+     * Return the set of preview language features used to declare the given element.
```
"Returns"

jonathan-gibbons reviewed (approved):
I have a mild queasiness about this new overloaded use of the word "Summary", since "summary tables" are normally the summary of the contents of a declaration, like fields and methods of a class.

That being said, the usage is primarily internal, and I have no overwhelmingly wonderful alternative, and (overloading aside) the term is accurate. 

So, OK for now. We can change it later if  we want to.

magicus reviewed (approved):
Build changes still good.

lahodaj commented:
I've merged the PR with the recent mainline, and I'd like to integrate sometime soon. Please let me know if there's any issue with that. Thanks!

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:
```diff
@@ -188,15 +192,34 @@ protected TypeElement getCurrentPageElement() {
 
     @Override @SuppressWarnings("preview")
     public void addClassSignature(String modifiers, Content classInfoTree) {
```
It seems less than ideal for this method to take a `String` and to then have to take it apart and reassemble it. It looks like it should be possible (and conceptually better) to change the signature to `List<String>` and to make the corresponding change to `utils.modifiersToString`, probably renaming it as `utils.modifiersToStrings` or something like that, and dropping the always-`true` argument for `trailingSpace`.  

But, the code is OK as is, and the suggestion is just for an enhancement, so is OK to defer it, if you would prefer.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java:
```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
```
(minor) now 2021

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java:
```diff
@@ -0,0 +1,303 @@
+ */
+public abstract class SummaryListWriter<L extends SummaryAPIListBuilder> extends SubWriterHolderWriter {
+
+    private String getAnchorName(SummaryElementKind kind) {
```
Heads-up: at some point this will conflict with another change in progress/review, to introduce a new type `HtmlId`  to use instead of `String` for ids.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java:
```diff
@@ -0,0 +1,303 @@
+     * @param headerKey table header key for the summary table
+     * @param contentTree the content tree to which the summary table will be added
+     */
+    protected void addSummaryAPI(SortedSet<Element> apiList, String id,
```
Heads-up, here's another occurrence of where there will be a conflict for ids.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/doclets.properties:
```diff
@@ -151,8 +151,6 @@ doclet.AnnotationInterfaces=Annotation Interfaces
 doclet.Exceptions=Exceptions
 doclet.Errors=Errors
 doclet.Classes=Classes
-doclet.packages=packages
```
I guess I'm mildly surprised to see all these items being removed...

lahodaj replied:
These were used from DeprecatedListWriter.getSummaryKey which is removed by this patch (and is unused in the current mainline, as far as I know).

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/SummaryAPIListBuilder.java:
```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 1998, 2020, Oracle and/or its affiliates. All rights reserved.
```
Year

jonathan-gibbons commented on test/langtools/jdk/javadoc/doclet/testPreview/TestPreview.java:
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
```
2021. I assume you will do a bulk update for affected files.

jonathan-gibbons reviewed (approved):
I've looked at all the files that were marked as changed since I last looked at them.

There's one suggested enhancement to reduce string bashing between `Utils` and `ClassWriterImpl` that could be done now or later.

There's a pending conflict with a PR of mine to change to use a new type `HtmlId` for HTML ids. This JEP12 work has been in progress for a while, and so it would be good to get it in before the `HtmlId` work, and I'll deal with the merge conflict in due course.

jonathan-gibbons commented:
+1

-- Jon

On 1/7/21 12:19 PM, Jan Lahoda wrote:
>
> I've merged the PR with the recent mainline, and I'd like to integrate 
> sometime soon. Please let me know if there's any issue with that. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://urldefense.com/v3/__https://github.com/openjdk/jdk/pull/703*issuecomment-756360085__;Iw!!GqivPVa7Brio!Oxxf6GpefmtstD2fCT8IKF4r-blOVCGWCibSjA4m4l24mI8j6j-RDEJAiBRtFfHCMqNb7Q$>, 
> or unsubscribe 
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AOUXBRTWQAZF7NEGTD3FSGDSYYJOPANCNFSM4STPEYJQ__;!!GqivPVa7Brio!Oxxf6GpefmtstD2fCT8IKF4r-blOVCGWCibSjA4m4l24mI8j6j-RDEJAiBRtFfFTnE3pGg$>.
>

lahodaj commented:
/integrate
